server_development/topics/auth-spi.adoc

```diff
@@ -108,199 +111,547 @@ Let's walk through the steps from when a client first redirects to keycloak to a
+    credentialData = new SecretQuestionCredentialData(question);
+    secretData = new SecretQuestionSecretData(answer);
+}
+
+public static SecretQuestionCredentialModel createSecretQuestion(String question, String answer) {
+    SecretQuestionCredentialModel credentialModel = new SecretQuestionCredentialModel(question, answer);
+    credentialModel.fillCredentialModelFields();
+    return credentialModel;
+}
+
+private void fillCredentialModelFields(){
+    try {
+        setCredentialData(JsonSerialization.writeValueAsString(credentialData));
+        setSecretData(JsonSerialization.writeValueAsString(secretData));
+        setType(TYPE);
+        setCreatedDate(Time.currentTimeMillis());
+    } catch (IOException e) {
+        throw new RuntimeException(e);
+    }
+}
+----
 
-To deploy this jar, just copy it to the providers directory. 
+==== Implementing a CredentialProvider
+
+As with all Providers, to allow {project_name} to generate the CredentialProvider, we require a CredentialProviderFactory. For this we create
+the SecretQuestionCredentialProviderFactory, whose `create` method will be called when a SecretQuestionCredentialProvider is asked for:
+
+[source,java]
+----
+public class SecretQuestionCredentialProviderFactory implements CredentialProviderFactory<SecretQuestionCredentialProvider> {
+
+    public static final String PROVIDER_ID =  "secret-question";
+
+    @Override
+    public String getId() {
+        return PROVIDER_ID;
+    }
+
+    @Override
+    public CredentialProvider create(KeycloakSession session) {
+        return new SecretQuestionCredentialProvider(session);
+    }
+}
+----
+
+The CredentialProvider interface takes a generic parameter that extends a CredentialModel. In our case we to use the SecretQuestionCredentialModel we created:
+
+[source,java]
+----
+public class SecretQuestionCredentialProvider implements CredentialProvider<SecretQuestionCredentialModel>, CredentialInputValidator {
+    private static final Logger logger = Logger.getLogger(SecretQuestionCredentialProvider.class);
+
+    protected KeycloakSession session;
+
+    public SecretQuestionCredentialProvider(KeycloakSession session) {
+        this.session = session;
+    }
+
+    private UserCredentialStore getCredentialStore() {
+        return session.userCredentialManager();
+    }
+----
+
+We also want to implement the CredentialInputValidator interface, as this allows {project_name} to know that this provider can also be used to validate a
+credential for an Authenticator. For the CredentialProvider interface, the first method that needs to be implemented is the `getType()` method. This will simply
+return the `SecretQuestionCredentialModel`'s TYPE String:
+
+[source,java]
+----
+@Override
+public String getType() {
+    return SecretQuestionCredentialModel.TYPE;
+}
+----
+
+The second method is to create a `SecretQuestionCredentialModel` from a `CredentialModel`. For this we simply call the existing static method
```

Comment:
For clarity, try to always add a noun after "this" for example
"For this method"